doc/developer/design/20240108_source_metrics_2.md

```diff
@@ -0,0 +1,208 @@
+
+#### Kafka
+
+`snapshot_total` can be be trivially exposed by exposing the snapshot frontier already tracked within its source operator,
+and summing across partitions. Similarly, `snapshot_progress` can be derived from the operator's output frontier.
+
+#### Postgres and MySQL
+
+`snapshot_total` will need to be calculated, in the unit of rows by performing `SELECT count(*)` on the tables that participate in the snapshot.
+Both the Postgres and MySQL implementations will be required to perform this query, per-table, during snapshotting. Note that `count(*)` is not
+guaranteed to be cheap on Postgres and MySQL. To avoid this, we will perform this query _concurrently_ with the beginning of
+snapshotting, allowing the user to see their source's progress before a percentage can be calculated.
+
+
+`snapshot_progress` will need to be exposed by the source operators themselves, as the data itself during snapshotting all has
+the same frontier. This means the operators will need to track and periodically report a frontier describing the progress they
+have made reading the snapshot. In practice, this is the number of rows they have read, per-table.
+
+#### Example Scenarios and Hydration Backpressure
+
+Below is a set of charts for that these new metrics and the derived percentage would look like in 2 difference scenarios.
+
+![snapshot-progress](./static/source_metrics_2/snapshot_progress.png)
+
+
+These charts make it clear that if we can _read_ from upstream faster than we can process that data, we may present a `snapshot_progress`
+that is artificially inflated. This is considered reasonable for this design because:
+
+- We will present the metrics as the _percentage of the snapshot Materialize has read_.
+- The console will present the user will the memory usage of the replica doing snapshotting, which makes it clear when this scenario is occurring
+(This is why memory is included on the above charts).
+- In general, once we have read the entire snapshot from upstream, we process the remaining data quite quickly. This is particularly true for non-UPSERT sources.
+- Categorically fixing this means implementing _Hydration Backpressure_, which effectively locks Materialize's _read-rate_ to its _process-rate_.
+This is considered [out of scope](#out-of-scope) for this design, but, _critically_: if and when _Hydration backpressure_ is implemented, these proposed
+metrics _become even more useful_, as opposed to becoming irrelevant.
+
+
+## Steady-state
+
+The second set of metrics this design document proposes describe the _rate_ at which upstream and Materialize are processing data.
+These metrics are designed to answer #2 in [the problem statement](#the-problem).
+
+```
+| `outstanding_total`    | [`uint8`] | The total number of outstanding values in the upstream source. The unit is source-specific.       |
+| `processed_total`      | [`uint8`] | The total number of upstream values Materialized has processed. Same unit as `outstanding_total`. |
```

Comment:
How will we maintain the ``outstanding_total` number across restarts? I wonder if it makes sense to just have workers report the currently oustanding work as a gauge without a total. What do you think?

Reply:
It will be a gauge! You are right that `_total` is misleading, as its not a counter; ill adjust the name

Reply:
changed to `_values`